File: js_nn/neuralnetwork/nn.js
function sigmoid(x){
  return 1/(1+ Math.exp(-x));
}
function dsigmoid(y){
  //return (sigmoid(x)*(1-sigmoid(x)));
  return y*(1-y);
}


class NeuralNetwork{
  constructor(a, b, c) {
    if (a instanceof NeuralNetwork) {
      this.input_nodes = a.input_nodes;
      this.hidden_nodes = a.hidden_nodes;
      this.output_nodes = a.output_nodes;

      this.weights_ih = a.weights_ih.copy();
      this.weights_ho = a.weights_ho.copy();

      this.bias_h = a.bias_h.copy();
      this.bias_o = a.bias_o.copy();
    } else {
      this.input_nodes = a;
      this.hidden_nodes = b;
      this.output_nodes = c;

      this.weights_ih = new Matrix(this.hidden_nodes, this.input_nodes);
      this.weights_ho = new Matrix(this.output_nodes, this.hidden_nodes);
      this.weights_ih.randomize();
      this.weights_ho.randomize();
      console.log(this.weights_ih);
      this.bias_h = new Matrix(this.hidden_nodes, 1);
      this.bias_o = new Matrix(this.output_nodes, 1);
      this.bias_h.randomize();
      this.bias_o.randomize();
      throw error;

    }

  }

  predict(input_array){
    //Generating hidden outputs
    let inputs = Matrix.fromArray(input_array);
    console.log("weights:", this.weights_ih)
    console.log("inputs: ", inputs)

    let hidden = Matrix.multiply(this.weights_ih, inputs);
    console.log("______")

    hidden.add(this.bias_h);
    hidden.map(sigmoid);
    //activation
    let output = Matrix.multiply(this.weights_ho, hidden);
    output.add(this.bias_o);
    output.map(sigmoid);


    return output.toArray();
  }
  train(input_array, target_array){
    //Generate hidden outputs
    let inputs = Matrix.fromArray(input_array);
    let hidden = Matrix.multiply(this.weights_ih, inputs);
    hidden.add(this.bias_h);
    //Activation
    hidden.map(sigmoid);
    //activation
    let outputs = Matrix.multiply(this.weights_ho, hidden);
    outputs.add(this.bias_o);
    outputs.map(sigmoid);


    // let outputs = this.feedforward(inputs);
    //Converting to Matrix object
    let targets = Matrix.fromArray(target_array);

    //Calculate error
    //Error = Targets - outputs
    let output_errors = Matrix.subtract(targets, outputs);

    //Gradient decent:
    let gradients = Matrix.map(outputs, dsigmoid);
    gradients.multiply(output_errors);
    gradients.multiply(this.learning_rate);


    //Calculatbe deltas
    let hidden_T = Matrix.transpose(hidden);
    let weight_ho_deltas = Matrix.multiply(outputs, hidden_T);

    //Adjust weights by deltas
    this.weights_ho.add(weight_ho_deltas);
    //Adjust the bias by its deltas
    this.bias_o.add(gradients);


    //Calculate the hidden layer errors
    let weights_ho_t = Matrix.transpose(this.weights_ho);
    let hidden_errors = Matrix.multiply(weights_ho_t, output_errors);
    //Calculate hidden gradient
    let hidden_gradient = Matrix.map(hidden,dsigmoid);
    hidden_gradient.multiply(hidden_errors);
    hidden_gradient.multiply(this.learning_rate);
    //Calculate input -> hidden deltas
    let inputs_T = Matrix.transpose(inputs);
    let weight_ih_deltas = Matrix.multiply(hidden_gradient,inputs_T);

    //Adjust weights by deltas
    this.weights_ih.add(weight_ih_deltas);
    //Adjust the bias by its deltas
    this.bias_h.add(hidden_gradient);
  }
  copy() {
    return new NeuralNetwork(this);
  }
  mutate(rate) {
    // This is how we adjust weights ever so slightly
    function mutate(x) {
      if (Math.random() < rate) {
        //Randomgaussian
        var offset = Math.random() * 0.5;
        // var offset = random(-0.1, 0.1);
        var newx = x + offset;
        return newx;
      } else {
        return x;
      }
    }
    this.weights_ih.map(mutate);
    this.weights_ho.map(mutate);
    this.bias_h.map(mutate);
    this.bias_o.map(mutate);
  }
  serialize() {
    return JSON.stringify(this);
  }
}

File: python_nn/nn/nn.py
import numpy as np
import random
class NeuralNetwork:
    def __init__(self,a,b,c):
        if isinstance(a,NeuralNetwork):
            self.input_nodes = a.input_nodes
            self.hidden_nodes = a.hidden_nodes
            self.output_nodes = a.output_nodes

            self.weights_ih = a.weights_ih.copy()
            self.weights_ho = a.weights_ho.copy()

            self.bias_h = a.bias_h.copy()
            self.bias_o = a.bias_o.copy()
        else:
            self.input_nodes = a
            self.hidden_nodes = b
            self.output_nodes = c

            #Random
            #self.weights_ih = np.array(self.hidden_nodes,self.input_nodes)
            #self.weights_ho = np.array(self.output_nodes,self.hidden_nodes)
            self.weights_ih = np.random.rand(self.hidden_nodes,self.input_nodes)*2-1
            self.weights_ho = np.random.rand(self.output_nodes,self.hidden_nodes)*2-1

            self.bias_h = np.random.rand(self.hidden_nodes,1)*2-1
            self.bias_o = np.random.rand(self.output_nodes,1) * 2 - 1

    def predict(self,input_array):
        sigFunc = np.vectorize(sigmoid)

        inputs = np.array(input_array)
        inputs.shape = (5,1) # Set correct shape of input array
        hidden = self.weights_ih.dot(inputs)

        hidden = np.add(hidden,self.bias_h)

        hidden = sigFunc(hidden)


        output = self.weights_ho.dot(hidden)
        output = output + self.bias_o
        output = sigFunc(output)
        return output.flatten()

    # def train(self,input_array,target_array):
    #     sigFunc = np.vectorize(sigmoid)
    #     dsigFunc = np.vectorize(dsigmoid)
    #
    #
    #     inputs = input_array
    #     hidden = self.weights_ih.dot(inputs)
    #
    #     hidden = sigFunc(hidden)
    #
    #     outputs = self.weights_ho.dot(hidden)
    #     outputs = outputs + self.bias_o
    #     outputs = sigFunc(outputs)
    #
    #     targets = target_array
    #     output_errors = targets - outputs
    #
    #     ## could be faulty above here
    #
    #     #Gradient decent
    #     gradients = dsigFunc(outputs)
    #     gradients = gradients.dot(output_errors)
    #     gradients = gradients.dot(self.learning_rate)
    #



    def copy_nn(self):
        new_class = NeuralNetwork(self)
        return new_class

    def mutate(self):
        mutateFunc = np.vectorize(mutateHelper)




def sigmoid(x):
    return 1/(1+np.exp(-x))
def dsigmoid(y):
    return y*(1-y)
def mutateHelper(x):
    #mutation rate
    rate = 0.1
    if(random.random() < rate):
        offset = random.random() * 0.5
        newX = x + offset
        return newX
    else:
        return x